python/ql/src/semmle/python/types/Exceptions.qll

```diff
@@ -274,9 +274,9 @@ class ExceptFlowNode extends ControlFlowNode {
         (
             this.getType().pointsTo(val, origin)
             or
-            exists(Value tup |
-                this.handledObject(val, ClassValue::tuple(), _) |
-                element_from_tuple(tup).pointsTo(val, origin)
+            exists(TupleValue tup |
+                this.handledObject(tup, ClassValue::tuple(), _) |
+                val = tup.getItem(_) and origin.pointsTo(val)
```

Comment:
I don't think `origin.pointsTo(val)` will do the right thing. Sure, at the actual origin, it _should_ point at `val`, but there might be several other control-flow nodes that point to that same value, and these shouldn't be counted as the origin.

Instead, perhaps something like `any(ControlFlowNode n).pointsTo(val, origin)` would work? That's a bit ugly, though.

Adding a `getOrigin` method to the `Value` type is starting to look more tempting. We already have this method on `ObjectInternal`, after all.

Reply:
I'll add `getOrigin` and rewrite the predicate to make it cleaner?

Reply:
Sounds good! Once you've added `getOrigin` to `Value`, you can also fix up `hasLocationInfo` to use this directly (instead of `(ObjectInternal).getOrigin`, which it does right now). I think that's the only place in that file where `getOrigin` is mentioned.

Reply:
Ok, `getOrigin` added. Should I also update `hasLocationInfo` here? I feel like because it's an orthogonal feature it might make sense to isolate it but maybe not?